code/modules/tgui_input/color.dm

```diff
@@ -0,0 +1,173 @@
+ * * autofocus - The bool that controls if this picker should grab window focus.
+ */
+/proc/tgui_color_picker_async(mob/user, message, title, default = "#000000", datum/callback/callback, timeout = 0, autofocus = TRUE)
+	if (!user)
+		user = usr
+	if (!istype(user))
+		if (istype(user, /client))
+			var/client/client = user
+			user = client.mob
+		else
+			return
+	// Client does NOT have tgui_input on: Returns regular input
+	if(!(user.client?.prefs?.toggles2 & PREFTOGGLE_2_TGUI_INPUT))
+		return input(user, message, title, default) as color|null
+	var/datum/tgui_color_picker/async/picker = new(user, message, title, default, callback, timeout, autofocus)
+	picker.ui_interact(user)
+
+/**
+ * # tgui_color_picker
+ *
+ * Datum used for instantiating and using a TGUI-controlled color picker.
+ */
+/datum/tgui_color_picker
+	/// The title of the TGUI window
+	var/title
+	/// The message to show the user
+	var/message
+	/// The default choice, used if there is an existing value
+	var/default
+	/// The color the user selected, null if no selection has been made
+	var/choice
+	/// The time at which the tgui_color_picker was created, for displaying timeout progress.
+	var/start_time
+	/// The lifespan of the tgui_color_picker, after which the window will close and delete itself.
+	var/timeout
+	/// The bool that controls if this modal should grab window focus
+	var/autofocus
+	/// Boolean field describing if the tgui_color_picker was closed by the user.
+	var/closed
+
+/datum/tgui_color_picker/New(mob/user, message, title, default, timeout, autofocus)
+	src.autofocus = autofocus
+	src.title = title
+	src.default = default
+	src.message = message
+	if (timeout)
+		src.timeout = timeout
+		start_time = world.time
+		QDEL_IN(src, timeout)
+
+/datum/tgui_color_picker/Destroy(force, ...)
+	SStgui.close_uis(src)
+	. = ..()
+
+/**
+ * Waits for a user's response to the tgui_color_picker's prompt before returning. Returns early if
+ * the window was closed by the user.
+ */
+/datum/tgui_color_picker/proc/wait()
+	while (!choice && !closed && !QDELETED(src))
+		stoplag(1)
+
+/datum/tgui_color_picker/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "ColorPickerModal")
+		ui.open()
+
+/datum/tgui_color_picker/ui_close(mob/user)
+	. = ..()
+	closed = TRUE
+
+/datum/tgui_color_picker/ui_state(mob/user)
+	return GLOB.always_state
+
+/datum/tgui_color_picker/ui_static_data(mob/user)
+	. = list()
+	.["autofocus"] = autofocus
+	.["large_buttons"] = !user.client?.prefs || (user.client.prefs.toggles2 & PREFTOGGLE_2_BIG_BUTTONS)
+	.["swapped_buttons"] = !user.client?.prefs || (user.client.prefs.toggles2 & PREFTOGGLE_2_SWITCHED_BUTTONS)
+	.["title"] = title
+	.["default_color"] = default
+	.["message"] = message
+
+/datum/tgui_color_picker/ui_data(mob/user)
+	. = list()
+	if(timeout)
+		.["timeout"] = CLAMP01((timeout - (world.time - start_time) - 1 SECONDS) / (timeout - 1 SECONDS))
```

Comment:
This won't work since the UI doesn't auto update, but with something as common and simple as colour picker it's peobably best if we dont send lots of updates and instead draw timeout based on a client side calculation

Reply:
It's copypasta from every other modal and TG

Reply:
Also nothing uses it so just ignore

Reply:
TG have auto-updating by default, we disabled it since 50% of UIs were constantly updating their ui_data despite not needing updates which was making TGUI one of the biggest things in the profiler.

Reply:
We have auto updating on by default, TG does not. That is our difference.

I'm saying it doesn't matter, since this code is copypasta from OTHER modals, and I would have to fix all modals that share this code if I fixed whatever the problem is here.

Reply:
Also `timeout` is used, QDEL_IN is called on the modal with it.

Reply:
I've fixed it now in my new commit